### PR TITLE
Arc 1282 add slauth headers verification

### PR DESCRIPTION
**Pull request by KAbakumovAtl**
Follow up on the previous PR.

Added a verification that SLAuth headers are set

Currently testing in staging

### CODE REVIEW

**Comment by KAbakumovAtl** [comment]
It works, tested in staging

File: src/routes/api/api-router.ts [hunk 1]
```diff
@@ -20,6 +20,36 @@ ApiRouter.use(urlencoded({ extended: false }));
...
+		}
+
+		req.log = req.log.child({slauth: {
+			mechanism,
+			issuer,
+			principal,
+			userGroup: req.get("X-Slauth-User-Groups"),
+			aaid: req.get("X-Slauth-User-Aaid"),
+			username: req.get("X-Slauth-User-Username")
+		}});
```
Comment (mboudreau) [inline src/routes/api/api-router.ts:44]: You'd probably want to set this before the conditional check

File: src/routes/api/api-router.ts [hunk 2]
```diff
@@ -20,6 +20,36 @@ ApiRouter.use(urlencoded({ extended: false }));
...
+
+ApiRouter.use(
+	async (req: Request, res: Response, next: NextFunction): Promise<void> => {
+		const mechanism = req.get("X-Slauth-Mechanism");
+		const issuer = req.get("X-Slauth-Issuer");
+		const principal = req.get("X-Slauth-Principal");
+
+		if(!mechanism || mechanism === "open") {
+			res.status(401).json({error: "Open access not allowed"});
+			return;
```
Comment (mboudreau) [inline src/routes/api/api-router.ts:34]: Should we maybe log this as a warn when someone's trying to authenticate without correct credentials?

Comment (KAbakumovAtl) [reply src/routes/api/api-router.ts:34]: Makes sense, will do

Comment (mboudreau) [reply src/routes/api/api-router.ts:34]: Also, I'm not sure I'm a fan of this conditional.  When you're doing a security check, you check for what you expect to continue, not check for what _not_ to continue.

Shouldn't we also check the other headers? is there something atlassian specific in the issues, principal, etc?  Shouldn't we also check the usergroup to make sure that whoever they are in atlassian has access to this API?

Comment (mboudreau) [reply src/routes/api/api-router.ts]: @KAbakumovAtl :arrow_up:

File: src/routes/api/api-router.ts [hunk 3]
```diff
@@ -20,6 +20,37 @@ ApiRouter.use(urlencoded({ extended: false }));
...
+			userGroup: req.get("X-Slauth-User-Groups"),
+			aaid: req.get("X-Slauth-User-Aaid"),
+			username: req.get("X-Slauth-User-Username")
+		}});
+
+		if(!mechanism || mechanism === "open") {
+			req.log.warn("Attempt to access Admin API without authentication")
+			res.status(401).json({error: "Open access not allowed"});
+			return;
+		}
```
Comment (mboudreau) [inline src/routes/api/api-router.ts:45]: Still need to fix this conditional.  You can't base your security on what not to expect.

File: src/routes/api/api-router.ts (hunk 2 above)
Comment (KAbakumovAtl) [reply src/routes/api/api-router.ts]: No, it is the responsibility of POCO

File: src/routes/api/api-router.ts (hunk 3 above)
Comment (KAbakumovAtl) [reply src/routes/api/api-router.ts:45]: Don't understand why. 
We are checking that headers present, we are not repeating POCO/SLAuth logic here. That's how it works in DSS and any other services which use SLAuth.

Our goal of checking this header to make sure that the SP is enabled (and that the endpoint is not public in the config). That's it.

Please refer to SLAuth docs go/slauth to get better understanding on how it works.

**Review by joshkay10 — APPROVED** [review]

**Pull request merged**